File: src/main/java/com/hoi4utils/clauzewitz/map/state/State.java
package main.java.com.hoi4utils.clauzewitz.map.state;

import main.java.com.hoi4utils.clauzewitz.HOIIVFile;
import main.java.com.hoi4utils.clauzewitz.localization.Localizable;
import main.java.com.hoi4utils.clauzewitz.code.ClausewitzDate;
import main.java.com.hoi4utils.clauzewitz.data.country.Country;
import main.java.com.hoi4utils.clauzewitz.data.country.CountryTag;
import main.java.com.hoi4utils.clauzewitz.data.country.CountryTagsManager;
import main.java.com.hoi4utils.clauzewitz.map.Owner;
import main.java.com.hoi4utils.clauzewitz.map.UndefinedStateIDException;
import main.java.com.hoi4utils.clauzewitz.map.buildings.Infrastructure;
import main.java.com.hoi4utils.clauzewitz.map.province.VictoryPoint;
import main.java.com.hoi4utils.clauzewitz.map.resources.Resources;
import main.kotlin.com.hoi4utils.clausewitz_parser.Node;
import main.kotlin.com.hoi4utils.clausewitz_parser.Parser;
import main.kotlin.com.hoi4utils.clausewitz_parser.ParserException;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.*;
import java.util.function.Function;

/**
 * Loads HOI4 State files, each instance represents a state as defined in "history/states"
 * Localizable: state name
 *
 * I apologize in advance.
 */
public class State implements InfrastructureData, Localizable, Iterable<State>, Comparable<State> {
	/* static */
	private static final ArrayList<State> states = new ArrayList<>();

	private File stateFile;
	private int stateID;
	private String name;
	private final Map<ClausewitzDate, Owner> owner;
	//! todo Finish state Category
	// private StateCategory stateCategory; 
	private Infrastructure stateInfrastructure;
	private Resources resourcesData;
	private List<VictoryPoint> victoryPoints;

	public State(File stateFile) {
		this(stateFile, true);
	}

	public State(File stateFile, boolean addToStatesList) {
		/* init */
		owner = new HashMap<>();

		this.stateFile = stateFile;
		this.name = stateFile.getName().replace(".txt", "");
		this.victoryPoints = new ArrayList<>();

		readStateFile(stateFile);

		// add to states list
		if (addToStatesList) {
			states.add(this);
		}
	}

	// todo simplify
	private void readStateFile(File stateFile) {
		int infrastructure = 0;
		int population = 0;
		int civilianFactories = 0;
		int militaryFactories = 0;
		int dockyards = 0;
		// ! todo something important
		// int navalPorts = 0; //has a province location
		int airfields = 0;

		/* parse state data */
		Parser stateParser = new Parser(stateFile);
		// Expression exp = stateParser.expressions();
		Node stateNode = null;
		try {
			stateNode = stateParser.parse().findFirst("state");
		} catch (ParserException e) {
			throw new RuntimeException(e);
		}

		// id
		if (stateNode.contains("id")) {
			stateID = stateNode.getValue("id").integer();
		} else {
			System.out.println(stateNode.value().asString());
			throw new UndefinedStateIDException(stateFile);
		}
		// population (manpower)
		if (stateNode.contains("manpower")) {
			population = stateNode.getValue("manpower").integer(); // todo after here etc.
		}
		// state category
		if (stateNode.contains("state_category")) {

		}

		/* buildings */
		if (stateNode.contains("history")) {
			Node historyNode = stateNode.findFirst("history");
			Node buildingsNode = null;
			if (historyNode.contains("buildings")) {
				buildingsNode = historyNode.findFirst("buildings");
			}
			// owner
			if (historyNode.contains("owner")) {
				// empty date constructor for default date
				owner.put(ClausewitzDate.of(), new Owner(new CountryTag(historyNode.getValue("owner").string())));
			} else {
				System.err.println("Warning: state owner not defined, " + stateFile.getName());
			}
			if (buildingsNode == null) {
				System.err.println("Warning: buildings (incl. infrastructure) not defined in state, " + stateFile.getName());
				stateInfrastructure = null;
			} else {
				// infrastructure
				if (buildingsNode.contains("infrastructure")) {
					infrastructure = buildingsNode.getValue("infrastructure").integer(); // todo after here etc.
				}
				// civilian factories
				if (buildingsNode.contains("industrial_complex")) {
					civilianFactories = buildingsNode.getValue("industrial_complex").integer(); // todo after here etc.
				}
				// military factories
				if (buildingsNode.contains("arms_factory")) {
					militaryFactories = buildingsNode.getValue("arms_factory").integer(); // todo after here etc.
				}
				// dockyards
				if (buildingsNode.contains("dockyard")) {
					dockyards = buildingsNode.getValue("dockyard").integer(); // todo after here etc.
				}
				// airfields
				if (buildingsNode.contains("air_base")) {
					airfields = buildingsNode.getValue("air_base").integer(); // todo after here etc.
				}
			}
			/* victory points */
			if (historyNode.contains("victory_points")) {
				Node victoryPointsNode = historyNode.findFirst("victory_points");
				for (Node vpNode : victoryPointsNode.toList()) {
					if (!vpNode.contains("province") || !vpNode.contains("value")) {
						System.out.println("Warning: invalid victory point node in state, " + stateFile.getName());
						continue;
					}
					var vp = VictoryPoint.of(vpNode.getValue("province").integer(),
							vpNode.getValue("value").integer());
					victoryPoints.add(vp);
				}
				//locHandler.addLocalization(victoryPoints, "VICTORY_POINTS_\\d+");
			}
		} else {
			System.out.println("Warning: history not defined in state, " + stateFile.getName());
			System.out.println(stateNode.value().asString());
		}

		// resources
		resourcesData = findStateResources(stateNode);

		// data record
		stateInfrastructure = new Infrastructure(population, infrastructure, civilianFactories, militaryFactories,
				dockyards, 0, airfields);
	}

	public static void read() {
		if (!HOIIVFile.mod_states_folder.exists() || !HOIIVFile.mod_states_folder.isDirectory()) {
			System.err.println("In State.java - " + HOIIVFile.mod_states_folder + " is not a directory, or etc.");
			return;
		}
		if (HOIIVFile.mod_states_folder.listFiles() == null || HOIIVFile.mod_states_folder.listFiles().length == 0) {
			System.out.println("No states found in " + HOIIVFile.mod_states_folder);
			return;
		}

		for (File stateFile : HOIIVFile.mod_states_folder.listFiles()) {
			if (stateFile.getName().endsWith(".txt"))
				new State(stateFile);
		}
	}

	public static ArrayList<State> list() {
		return states;
	}

	public static ArrayList<State> ownedStatesOfCountry(Country country) {
		return ownedStatesOfCountry(country.countryTag());
	}

	public static ArrayList<State> ownedStatesOfCountry(CountryTag tag) {
		ArrayList<State> countryStates = new ArrayList<>();

		for (State state : states) {
			Owner owner = state.owner.get(ClausewitzDate.defaulty());
			if (owner != null) {
				if (owner.isCountry(tag)) {
					countryStates.add(state);
				}
			} else {
				System.out.println("Undefined owner for state: " + state);
			}
		}

		return countryStates;
	}

	public static Infrastructure infrastructureOfStates(ArrayList<State> states) {
		int infrastructure = 0;
		int population = 0;
		int civilianFactories = 0;
		int militaryFactories = 0;
		int dockyards = 0;
		int airfields = 0;

		for (State state : states) {
			Infrastructure stateData = state.getStateInfrastructure();
			infrastructure += stateData.infrastructure();
			population += stateData.population();
			civilianFactories += stateData.civilianFactories();
			militaryFactories += stateData.militaryFactories();
			dockyards += stateData.navalDockyards();
			airfields += stateData.airfields();
		}

		return new Infrastructure(population, infrastructure, civilianFactories, militaryFactories, dockyards, 0,
				airfields);
	}

	// todo this is called and ran lots of times, optimize?
	public static Resources resourcesOfStates(ArrayList<State> states) {
		Resources resourcesOfStates = new Resources();
//		int aluminum = 0;
//		int chromium = 0;
//		int oil = 0;
//		int rubber = 0;
//		int steel = 0;
//		int tungsten = 0;

		for (State state : states) {
			Resources resources = state.getResources();
			resourcesOfStates.add(resources);
		}

		//return new Resources(aluminum, chromium, oil, rubber, steel, tungsten);
		System.out.println(resourcesOfStates.get("aluminum").amt());
		return resourcesOfStates;
	}

	public static Resources resourcesOfStates() {
		return resourcesOfStates(states);
	}

	public static int numStates(CountryTag country) {
		return ownedStatesOfCountry(country).size();
	}

	public static State get(String state_name) {
		state_name = state_name.trim();
		for (State state : states) {
			if (state.name.equals(state_name)) {
				return state;
			}
		}
		return null;
	}

	public static State get(File file) {
		for (State state : states) {
			if (state.stateFile.equals(file)) {
				return state;
			}
		}

		return null;
	}

	/**
	 * If the state represented by file is not in the list of states, creates the
	 * new state.
	 * If the state already exists, overwrites the state.
	 * 
	 * @param file state file
	 */
	public static void readState(File file) {
		State tempState = new State(file, false);
		if (tempState.stateID < 1) {
			System.err.println("Error: Invalid state id for state " + tempState);
			return;
		}

		for (State state : states) {
			if (state.stateID == tempState.stateID) {
				states.remove(state);
				states.add(tempState);
				System.out.println("Modified state " + tempState);
				return;
			}
		}

		// if state did not exist already in states
		states.add(tempState);
		System.out.println("Added state " + tempState);
	}

	/**
	 * If the state represented by the file exists in states list, removes the state
	 * from the states list
	 * 
	 * @param file state file
	 */
	public static void deleteState(File file) {
		State tempState = new State(file, false);
		if (tempState.stateID < 1) {
			System.err.println("Error: Invalid state id for state " + tempState);
			return;
		}
		for (State state : states) {
			if (state.stateID == tempState.stateID) {
				states.remove(state);
				System.out.println("Removed state " + tempState);
				return;
			}
		}

		System.out.println("Tried to delete state represented by file: " + "\n\t" + file + "\n\t"
				+ "but state not found in states list");
	}

	public static State get(int id) {
		return states.stream().filter(state -> state.stateID == id).findFirst().orElse(null);
	}

	public static List<Function<State, ?>> getStateDataFunctions(boolean resourcePercentages) {
		List<Function<State, ?>> dataFunctions = new ArrayList<>(17);         // for optimization, limited number of data functions.

		dataFunctions.add(State::id);
		dataFunctions.add(c -> c.stateInfrastructure.population());
		dataFunctions.add(c -> c.stateInfrastructure.civMilRatio());
		dataFunctions.add(c -> c.stateInfrastructure.militaryFactories());
		dataFunctions.add(c -> c.stateInfrastructure.navalDockyards());
		dataFunctions.add(c -> c.stateInfrastructure.airfields());
		dataFunctions.add(c -> c.stateInfrastructure.civMilRatio());
		dataFunctions.add(c -> c.stateInfrastructure.popPerFactoryRatio());
		dataFunctions.add(c -> c.stateInfrastructure.popPerCivRatio());
		dataFunctions.add(c -> c.stateInfrastructure.popPerMilRatio());
		dataFunctions.add(c -> c.stateInfrastructure.popAirportCapacityRatio());
		/* todo better way to do this obv! plz fix :(
		with (wrapper function that returns either or depndent on resourcesPerfcentages boolean value ofc */
		// also if we're gonna have different resources able to load in down the line... it'll break this.
		if (resourcePercentages) {
			dataFunctions.add(s -> s.getResources().get("aluminum").amt());
			dataFunctions.add(s -> s.getResources().get("chromium").amt());
			dataFunctions.add(s -> s.getResources().get("oil").amt());
			dataFunctions.add(s -> s.getResources().get("rubber").amt());
			dataFunctions.add(s -> s.getResources().get("steel").amt());
			dataFunctions.add(s -> s.getResources().get("tungsten").amt());
		} else {
			dataFunctions.add(s -> s.getResources().get("aluminum").percentOfGlobal());
			dataFunctions.add(s -> s.getResources().get("chromium").percentOfGlobal());
			dataFunctions.add(s -> s.getResources().get("oil").percentOfGlobal());
			dataFunctions.add(s -> s.getResources().get("rubber").percentOfGlobal());
			dataFunctions.add(s -> s.getResources().get("steel").percentOfGlobal());
			dataFunctions.add(s -> s.getResources().get("tungsten").percentOfGlobal());
		}

		return dataFunctions;
	}


	public Resources findStateResources(Node stateNode) {
		int aluminum = 0;
		int chromium = 0;
		int oil = 0;
		int rubber = 0;
		int steel = 0;
		int tungsten = 0;

		if (!stateNode.contains("resources")) {
			return new Resources(aluminum, chromium, oil, rubber, steel, tungsten);
		}

		/* resources */
		Node resourcesNode = stateNode.findFirst("resources");
		// aluminum (aluminium bri'ish spelling)
		if (resourcesNode.contains("aluminium")) {        // ! todo always null
			aluminum = (int) resourcesNode.getValue("aluminium").rational();
		}
		// chromium
		if (resourcesNode.contains("chromium")) {
			chromium = (int) resourcesNode.getValue("chromium").rational();
		}
		// rubber
		if (resourcesNode.contains("rubber")) {
			rubber = (int) resourcesNode.getValue("rubber").rational();
		}
		// oil
		if (resourcesNode.contains("oil")) {
			oil = (int) resourcesNode.getValue("oil").rational();
		}
		// steel
		if (resourcesNode.contains("steel")) {
			steel = (int) resourcesNode.getValue("steel").rational();
		}
		// tungsten
		if (resourcesNode.contains("tungsten")) {
			tungsten = (int) resourcesNode.getValue("tungsten").rational();
		}

		return new Resources(aluminum, chromium, oil, rubber, steel, tungsten);
	}

	public static List<Infrastructure> infrastructureOfCountries() {
		List<CountryTag> countryList = CountryTagsManager.getCountryTags();
		List<Infrastructure> countriesInfrastructureList = new ArrayList<>();

		for (CountryTag tag : countryList) {
			countriesInfrastructureList.add(infrastructureOfCountry(tag));
		}

		return countriesInfrastructureList;
	}

	private static Infrastructure infrastructureOfCountry(CountryTag tag) {
		return infrastructureOfStates(ownedStatesOfCountry(tag));
	}

	// ! todo test if working
	public static List<Resources> resourcesOfCountries() {
		List<CountryTag> countryList = CountryTagsManager.getCountryTags();
		List<Resources> countriesResourcesList = new ArrayList<>();

		for (CountryTag tag : countryList) {
			countriesResourcesList.add(resourcesOfCountry(tag));
		}

		return countriesResourcesList;
	}

	public static Resources resourcesOfCountry(CountryTag tag) {
		return resourcesOfStates(ownedStatesOfCountry(tag));
	}

	public Infrastructure getStateInfrastructure() {
		return stateInfrastructure;
	}

	public Resources getResources() {
		return resourcesData;
	}

	public String toString() {
		return name;
	}

	protected static boolean usefulData(String data) {
		if (!data.isEmpty()) {
			if (data.trim().charAt(0) == '#') {
				return false;
			} else {
				return true;
			}
		} else {
			return false;
		}
	}

	public File getFile() {
		return stateFile;
	}

	public Infrastructure getInfrastructureRecord() {
		return stateInfrastructure;
	}

	public int id() {
		return stateID;
	}

	@NotNull
	@Override
	public Iterator<State> iterator() {
		return states.iterator();
	}

	@Override
	public int compareTo(@NotNull State o) {
		return Integer.compare(stateID, o.stateID);
	}

	@Override
	public @NotNull Map<Property, String> getLocalizableProperties() {
		return Map.of(Property.NAME, name);
	}

	@Override
	public @NotNull Collection<? extends Localizable> getLocalizableGroup() {
		return states;
	}
}

File: src/main/scala/com/hoi4utils/clausewitz/map/state/State.scala
package com.hoi4utils.clausewitz.map.state

import com.hoi4utils.clausewitz.code.ClausewitzDate
import com.hoi4utils.clausewitz.data.country.{Country, CountryTag, CountryTagsManager}
import com.hoi4utils.clausewitz.localization.*
import com.hoi4utils.clausewitz.map.{Owner, UndefinedStateIDException}
import com.hoi4utils.clausewitz.map.buildings.Infrastructure
import com.hoi4utils.clausewitz.map.province.Province
import com.hoi4utils.clausewitz.map.state.State.History
import com.hoi4utils.clausewitz.script.*
import com.hoi4utils.clausewitz.{HOIIVFiles, HOIIVUtils}
import com.hoi4utils.clausewitz_parser.*
import javafx.collections.{FXCollections, ObservableList}
import org.apache.logging.log4j.{LogManager, Logger}
import org.jetbrains.annotations.NotNull

import java.io.File
import scala.collection.mutable
import scala.collection.mutable.{ListBuffer, Map}
import scala.jdk.CollectionConverters.*
import scala.jdk.javaapi.CollectionConverters
import scala.util.{Failure, Success, Try}

/**
 * Represents a state in HOI4
 * @param stateFile state file
 * @param addToStatesList if true, adds the state to the list of states
 */
class State(addToStatesList: Boolean) extends StructuredPDX("state") with InfrastructureData with Localizable with Iterable[Province] with Comparable[State] {
  private val LOGGER: Logger = LogManager.getLogger(getClass)

  final val stateID = new IntPDX("id")
  final val name = new StringPDX("name")
  final val resources = new CollectionPDX[Resource](Resource(), "resources") {
    override def loadPDX(expr: Node): Unit = {
      try {
        super.loadPDX(expr)
      } catch {
        case e: Exception =>
          this.LOGGER.error(s"Error loading resources for state ${stateID.value.getOrElse("unknown")}: ${e.getMessage}")
      }
    }

    override def getPDXTypeName: String = "Resources"
  }
  final val history = new History
  final val provinces = {
    val loadNewProvince = () => {val p = new Province(); Province.add(p); p}
    new ListPDX[Province](loadNewProvince, "provinces")
  }
  final val manpower = new IntPDX("manpower")
  final val buildingsMaxLevelFactor = new DoublePDX("buildings_max_level_factor")
  final val state_category = new StateCategory("state_category")
  final val local_supplies = new DoublePDX("local_supplies")

  private var _stateFile: Option[File] = None

  /* init */
  if (addToStatesList) State.add(this)

  def this(addToStatesList: Boolean, file: File) = {
    this(addToStatesList)
    loadPDX(file)
    setFile(file)
  }

  /**
   * @inheritdoc
   */
  override protected def childScripts: mutable.Iterable[PDXScript[?]] = {
    ListBuffer(stateID, name, resources, history, provinces, manpower, buildingsMaxLevelFactor,
      state_category, local_supplies)
  }

//  private def parseStateNode(stateFile: File): Option[Node] = {
//    val stateParser = new Parser(stateFile)
//    Try(stateParser.parse.find("state").get) match {
//      case Success(node) => Some(node)
//      case Failure(e: ParserException) => throw new RuntimeException(e)
//      case Failure(_) => None
//    }
//  }
//
//  private def extractStateID(stateNode: Node, stateFile: File): Int = {
//    if (stateNode.contains("id")) {
//      stateNode.getValue("id").integer
//    } else {
//      LOGGER.error(s"State ID not found in file: ${stateFile.getName}, ${stateNode.$}")
//      throw new UndefinedStateIDException(stateFile)
//    }
//  }

//  // todo breh nah
//  private def extractPopulation(stateNode: Node): Int = {
//    stateNode.find("manpower").map(_.value match {
//      case Some(v) => v match {
//        case i: Int => i
//        case s: String => s.toIntOption.getOrElse(0) // Convert string to Int safely
//        case _ => 0
//      }
//      case None => 0
//    }).getOrElse(0)
//  }
//
//
//  private def extractStateCategory(stateNode: Node): Unit = {
//    if (stateNode.contains("state_category")) {
//      // TODO: Implement state category logic
//    }
//  }
//
//  private def extractOwner(historyNode: Node, stateFile: File): Unit = {
//    if (historyNode.contains("owner")) {
//      historyNode.find("owner").get.$string match {
//        case Some(ownerTag) => owner.put(ClausewitzDate.of, new Owner(new CountryTag(ownerTag)))
//        case None => LOGGER.error(s"Warning: state owner not defined in ${stateFile.getName}")
//      }
//    } else {
//      LOGGER.error(s"Warning: state owner not defined in ${stateFile.getName}")
//    }
//  }

//  private def extractVictoryPoints(historyNode: Node, stateFile: File): Unit = historyNode.find("victory_points") match {
//    case Some(victoryPointsNode) =>
//      val vpl = CollectionConverters.asJava(victoryPointsNode.toList)
//      if (vpl.size == 2) {
////        victoryPoints.addOne(VictoryPoint.of(vpl.get(0).identifier.toInt, vpl.get(1).identifier.toInt)) //TODO
//      } else {
//        LOGGER.warn(s"Invalid victory point node in state: ${stateFile.getName}")
//      }
//    case None =>  LOGGER.info(s"Victory points not defined in state: ${stateFile.getName}")
//  }


//  private def findStateResources(stateNode: Node): Resources = {
//    if (!stateNode.contains("resources")) return Resources()
//
//    val resourcesNode = stateNode.find("resources").orNull
//
//    // aluminum (aluminium bri'ish spelling)
//    val aluminum: Int = resourcesNode.find("aluminium") match {
//      case Some(al) => al.$intOrElse(0)
//      case None => 0
//    }
//    val chromium: Int = resourcesNode.find("chromium") match {
//      case Some(ch) => ch.$intOrElse(0)
//      case None => 0
//    }
//    val oil: Int = resourcesNode.find("oil") match {
//      case Some(o) => o.$intOrElse(0)
//      case None => 0
//    }
//    val rubber: Int = resourcesNode.find("rubber") match {
//      case Some(r) => r.$intOrElse(0)
//      case None => 0
//    }
//    val steel: Int = resourcesNode.find("steel") match {
//      case Some(s) => s.$intOrElse(0)
//      case None => 0
//    }
//    val tungsten: Int = resourcesNode.find("tungsten") match {
//      case Some(t) => t.$intOrElse(0)
//      case None => 0
//    }
//
//    new Resources(aluminum, chromium, oil, rubber, steel, tungsten)
//  }

  def getStateInfrastructure: Infrastructure = {
    new Infrastructure(population, infrastructure, civilianFactories, militaryFactories, navalDockyards, navalPorts, airfields)
  }

  def listResources: List[Resource] = {
    resources.toList
  }

  override def toString: String = {
    name.value match {
      case Some(n) => n
      case None => stateID.value match {
        case Some(i) => i.toString
        case None => "[Unknown]"
      }
    }
  }

  def stateFile: Option[File] = _stateFile

  def setFile(file: File): Unit = {
    _stateFile = Some(file)
  }

  def resourceAmount(name: String): Double = resources.find(_.pdxTypeIdentifier.equals(name)) match {
    case Some(r) => r.getOrElse(0)
    case None => 0.0
  }
  
  def owner(date: ClausewitzDate): Option[CountryTag] = {
    history.owner.value match {
      case Some(owner) => Some(owner)
      case None => None
    }
  }

  def population: Int = manpower.getOrElse(1)

  def infrastructure: Int = history.buildings.infrastructure.getOrElse(0)

  def civilianFactories: Int = history.buildings.civilianFactories.getOrElse(0)

  def militaryFactories: Int = history.buildings.militaryFactories.getOrElse(0)

  def navalDockyards: Int = history.buildings.navalDockyards.getOrElse(0)

  def navalPorts: Int = history.buildings.navalDockyards.getOrElse(0)

  def airfields: Int = history.buildings.airBase.getOrElse(0)

  def civMilRatio: Double = {
    if (civilianFactories == 0) 0.0
    else militaryFactories.toDouble / civilianFactories
  }

  def popPerFactoryRatio: Double = {
    if (civilianFactories == 0) 0.0
    else population.toDouble / civilianFactories
  }

  def popPerCivRatio: Double = {
    if (civilianFactories == 0) 0.0
    else population.toDouble / civilianFactories
  }

  def popPerMilRatio: Double = {
    if (militaryFactories == 0) 0.0
    else population.toDouble / militaryFactories
  }

  def popAirportCapacityRatio: Double = {
    if (airfields == 0) 0.0
    else population.toDouble / airfields
  }

  def infrastructureMaxLevel: Int = {
    if (buildingsMaxLevelFactor.getOrElse(0) == 0) 0
    else (infrastructure * buildingsMaxLevelFactor.getOrElse(0)).toInt
  }

  def resource(name: String): Resource = resources.find(_.pdxTypeIdentifier.equals(name)) match {
    case Some(r) => r
    case None => new Resource(name, 0)
  }

  override def getInfrastructureRecord: Infrastructure = getStateInfrastructure

  override def compareTo(@NotNull o: State): Int = stateID.compareTo(o.stateID) match {
    case Some(v) => v
    case None => throw new NullPointerException("State has no ID")
  }

  @NotNull override def getLocalizableProperties: mutable.Map[Property, String] = {
    val id = this.stateID.value match {
      case Some(i) => i.toString
      case None => ""
    }
    mutable.Map(Property.NAME -> id)
  }

  @NotNull override def getLocalizableGroup: Iterable[? <: Localizable] = State.states

  override def iterator: Iterator[Province] = provinces.iterator
}

/**
 * Loads HOI4 State files, each instance represents a state as defined in "history/states"
 * Localizable: state name
 *
 * I apologize in advance.
 */
object State extends Iterable[State] {
  private val LOGGER: Logger = LogManager.getLogger(getClass)
  /* static */
  private val states = new ListBuffer[State]

  def get(file: File): Option[State] = {
    if (file == null) return None
    if (!states.exists(_.stateFile.contains(file))) new State(true, file)
    states.find(_.stateFile.contains(file))
  }

  def observeStates: ObservableList[State] = {
    FXCollections.observableArrayList(CollectionConverters.asJava(states))
  }

  /**
   * Creates States from reading files
   */
  def read(): Boolean = {
    if (!HOIIVFiles.Mod.states_folder.exists || !HOIIVFiles.Mod.states_folder.isDirectory) {
      LOGGER.fatal(s"In State.java - ${HOIIVFiles.Mod.states_folder} is not a directory, or it does not exist.")
      false
    } else if (HOIIVFiles.Mod.states_folder.listFiles == null || HOIIVFiles.Mod.states_folder.listFiles.isEmpty) {
      LOGGER.fatal(s"No states found in ${HOIIVFiles.Mod.states_folder}")
      false
    } else {
      LOGGER.info(s"Reading states from ${HOIIVFiles.Mod.states_folder}")

      HOIIVFiles.Mod.states_folder.listFiles().filter(_.getName.endsWith(".txt")).foreach { f =>
        new State(true, f)
      }
      true
    }
  }

  def clear(): Unit = {
    states.clear()
  }

  def add(state: State): Iterable[State] = {
    states += state
    states
  }

  def list: List[State] = states.toList

  def get(id: Int): Option[State] ={
    states.find(_.stateID @== id)
  }

  def get(state_name: String): Option[State] = {
    states.find(_.name @== state_name)
  }

  def ownedStatesOfCountry(country: Country): ListBuffer[State] = ownedStatesOfCountry(country.countryTag)

  def ownedStatesOfCountry(tag: CountryTag): ListBuffer[State] = {
    states filter(state => state.owner(ClausewitzDate.defaulty).exists(_.equals(tag)))
  }

  def infrastructureOfStates(states: ListBuffer[State]): Infrastructure = {
    var infrastructure = 0
    var population = 0
    var civilianFactories = 0
    var militaryFactories = 0
    var dockyards = 0
    var airfields = 0
    for (state <- states) {
      val stateData = state.getStateInfrastructure
      infrastructure += stateData.infrastructure
      population += stateData.population
      civilianFactories += stateData.civilianFactories
      militaryFactories += stateData.militaryFactories
      dockyards += stateData.navalDockyards
      airfields += stateData.airfields
    }
    new Infrastructure(population, infrastructure, civilianFactories, militaryFactories, dockyards, 0, airfields)
  }

  // todo this is called and ran lots of times, optimize?
  def resourcesOfStates(states: ListBuffer[State]): List[Resource] = {
    //    val resourcesOfStates = new Resources
    //    for (state <- states) {
    //      val resources = state.getResources
    //      resourcesOfStates.add(resources)
    //    }
    //    //return new Resources(aluminum, chromium, oil, rubber, steel, tungsten);
    //    System.out.println(resourcesOfStates.get("aluminum").amt)
    //    resourcesOfStates
    if (states.isEmpty) Resource.newList()
    else {
      states
        .flatMap(_.listResources).groupBy(_.pdxTypeIdentifier).values.map { resources =>
        new Resource(resources.head.pdxTypeIdentifier, resources.map(_.getOrElse(0)).sum)
      }.toList
    }
  } 

  def resourcesOfStates: List[Resource] = resourcesOfStates(states)

  def numStates(country: CountryTag): Int = ownedStatesOfCountry(country).size

  /**
   * If the state represented by file is not in the list of states, creates the
   * new state.
   * If the state already exists, overwrites the state.
   *
   * @param file state file
   */
  def readState(file: File): Boolean = {
    if (file == null || !file.exists || file.isDirectory) {
      LOGGER.fatal(s"In State.java - ${file} is a directory, or it does not exist.")
      false
    } else {
      LOGGER.info(s"Reading state from ${file}")
      new State(true, file)
      true
    }
  }

  /**
   * If the state represented by the file exists in states list, removes the state
   * from the states list
   *
   * @param file state file
   */
  def removeState(file: File): Boolean = {
    val tempState = new State(false)
    for (state <- states) {
      if (state.stateID == tempState.stateID) {
        states -= state
        LOGGER.debug("Removed state " + tempState)
        return true
      }
    }
    false
  }

  def getDataFunctions(resourcePercentages: Boolean): Iterable[State => ?] = {
    val dataFunctions = ListBuffer[State => ?]()

    dataFunctions += (s => s.stateID)
    dataFunctions += (s => s.population)
    dataFunctions += (s => s.civMilRatio)
    dataFunctions += (s => s.militaryFactories)
    dataFunctions += (s => s.navalDockyards)
    dataFunctions += (s => s.airfields)
    dataFunctions += (s => s.civMilRatio)
    dataFunctions += (s => s.popPerFactoryRatio)
    dataFunctions += (s => s.popPerCivRatio)
    dataFunctions += (s => s.popPerMilRatio)
    dataFunctions += (s => s.popAirportCapacityRatio)
    /* todo better way to do this obv! plz fix :(
        with (wrapper function that returns either or depndent on resourcesPerfcentages boolean value ofc */
    // also if we're gonna have different resources able to load in down the line... it'll break this.
    if (resourcePercentages) {
      dataFunctions += (s => s.resourceAmount("aluminum"))
      dataFunctions += (s => s.resourceAmount("chromium"))
      dataFunctions += (s => s.resourceAmount("oil"))
      dataFunctions += (s => s.resourceAmount("rubber"))
      dataFunctions += (s => s.resourceAmount("steel"))
      dataFunctions += (s => s.resourceAmount("tungsten"))
    }
    else {
      val globalResources = states.flatMap(_.listResources)
      dataFunctions += (s => s.resource("aluminum").percentOfGlobal(globalResources))
      dataFunctions += (s => s.resource("chromium").percentOfGlobal(globalResources))
      dataFunctions += (s => s.resource("oil").percentOfGlobal(globalResources))
      dataFunctions += (s => s.resource("rubber").percentOfGlobal(globalResources))
      dataFunctions += (s => s.resource("steel").percentOfGlobal(globalResources))
      dataFunctions += (s => s.resource("tungsten").percentOfGlobal(globalResources))
    }
    dataFunctions
  }

  def infrastructureOfCountries: ListBuffer[Infrastructure] = {
    val countryList = CountryTagsManager.getCountryTags.asScala
    val countriesInfrastructureList = new ListBuffer[Infrastructure]
    for (tag <- countryList) {
      countriesInfrastructureList.addOne(infrastructureOfCountry(tag))
    }
    countriesInfrastructureList
  }

  private def infrastructureOfCountry(tag: CountryTag) = infrastructureOfStates(ownedStatesOfCountry(tag))

  // ! todo test if working
  def resourcesOfCountries: List[List[Resource]] = {
    val countryList = CountryTagsManager.getCountryTags.asScala
    val countriesResourcesList = new ListBuffer[List[Resource]]
    for (tag <- countryList) {
      countriesResourcesList.addOne(resourcesOfCountry(tag))
    }
    countriesResourcesList.toList
  }

  def resourcesOfCountry(tag: CountryTag): List[Resource] = resourcesOfStates(ownedStatesOfCountry(tag))

  protected def usefulData(data: String): Boolean = if (data.nonEmpty) if (data.trim.charAt(0) == '#') false
  else true
  else false

  @NotNull override def iterator: Iterator[State] = states.iterator

  class History extends StructuredPDX("history") {
    final val owner = new ReferencePDX[CountryTag](() => CountryTag.toList, tag => Some(tag.get), "owner")
    final val buildings = new BuildingsPDX
//    final MultiPDX[VictoryPoint] victoryPoints = new MultiPDX(None, Some())
    //final val addCoreOf = new MultiReferencePDX[CountryTag](() => CountryTag.toList, tag => Some(tag.get), List("add_core_of")) // TODO

    def this(node: Node) = {
      this()
      loadPDX(node)
    }

    /**
     * @inheritdoc
     */
    override protected def childScripts: mutable.Iterable[PDXScript[?]] = {
      ListBuffer(owner, buildings)
    }
  }

  class BuildingsPDX extends StructuredPDX("buildings") {
    final val infrastructure = new IntPDX("infrastructure")
    final val civilianFactories = new IntPDX("industrial_complex")
    final val militaryFactories = new IntPDX("arms_factory")
    final val navalDockyards = new IntPDX("naval_base")
    final val airBase = new IntPDX("air_base")

    def this(node: Node) = {
      this()
      loadPDX(node)
    }

    /**
     * @inheritdoc
     */
    override protected def childScripts: mutable.Iterable[PDXScript[?]] = {
      ListBuffer(infrastructure, civilianFactories, militaryFactories, navalDockyards, airBase)
    }
  }
}